058-kraft-upgrades-and-downgrades.md

```diff
@@ -0,0 +1,221 @@
+
+#### Configuring the metadata version
+
+In ZooKeeper based clusters, the `inter.broker.protocol.version` is stored in the Kafka configuration (`Kafka.spec.kafka.config`).
+The `metadata.version` is not part of the Kafka configuration and cannot be managed declaratively.
+To allow users configure it in our Kafka custom resource, a new String-type field named `metadataVersion` will be added to `Kafka.spec.kafka`.
+This field will be validated by the operator to contain a valid Kafka metadata version.
+
+Users can use this new field to configure the metadata version.
+The default value of this field when not set will be the metadata version corresponding to the current Kafka version.
+The default version corresponding to given Kafka version will be stored in `kafka-versions.yaml` in the same way we today store the default `inter.broker.protocol.version` and `log.message.format.version`.
+
+The operator will use this field to set the metadata version in the Kafka cluster.
+In the initial deployment of the Kafka cluster, it will be set using the `kafka-storage.sh` tool.
+In existing Kafka clusters, it will be queried and changed using the Kafka Admin API.
+
+The current metadata version will be tracked in the `.status` section of the Kafka CR in `.status.kafkaMetadataVersion`.
+This field will be used for validations and to track if upgrade / downgrade can be executed.
+
+##### Upgrading metadata version
+
+When upgrading the metadata version, the operator will validate the desired version and check if it is expected to work with the used Kafka version (e.g. is not higher than `metadata.version` supported by the Kafka version).
+If it passes the validation, Kafka Admin API will be used to upgrade the metadata.
+If the validation fails, the existing metadata version will be used and the error will be raised in log and in the `.status` section of the Kafka CR.
+
+##### Downgrading metadata version
+
+When downgrading the `metadata.version` in Kafka, the operator will always try to do a _safe_ downgrade only.
+And if the safe downgrade fails, it will report an error (in log and in `.status` section of the Kafka CR) and continue reconciliation without changing the metadata version.
+
+In case the safe downgrade is not supported, an _unsafe_ downgrade can be attempted (as explained in one of the earlier sections, the _unsafe_ downgrade is not implemented in current Kafka versions, but it is planned to be supported in the future).
+Unsafe upgrade might result in metadata loss that might cause various problems to the Kafka cluster.
+Users who decide they want to do the unsafe downgrade can do so manually:
+* Pause the reconciliation of the Kafka CR
+* Update the Kafka CR to the updated metadata version (to avoid the operator upgrading it back immediately after the reconciliation is unpaused)
+* Use the Kafka Admin API or CLI tools to do the unsafe downgrade
```

Comment:
I think the `safe` part means, that the metadata version between versions used during downgrade are compatible. If not user can still do it, but it could break something, correct @scholzj ?

Reply:
Yes. The safe downgrade will be done by the operator as it should guarantee that the stuff will keep working. The unsafe downgrade has to be done by the users and might cause issues and metadata loss.

Reply:
Okay. Got it.